### PR TITLE
feature/225_add_cache_oauth2_hive_auth_provider

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #225 
* Unit tests passed:
```
$ mvn test -Duser=frb -Dtoken=xxxxx
...
...
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.655s
[INFO] Finished at: Thu Mar 02 09:08:01 CET 2017
[INFO] Final Memory: 12M/223M
[INFO] ------------------------------------------------------------------------
```

* e2e tests passed:
```
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: Hive configuration read
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: com.telefonica.iot.idm.endpoint=https://account.lab.fiware.org
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.enable.doAs=true
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.authentication=CUSTOM
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.custom.authentication.class=com.telefonica.iot.cosmos.hive.authprovider.OAuth2AuthenticationProviderImpl
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hadoop.proxyuser.hive.groups=*
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hadoop.proxyuser.hive.hosts=*
17/03/02 07:49:42 INFO authprovider.HttpClientFactory: Setting max total connections (500)
17/03/02 07:49:42 INFO authprovider.HttpClientFactory: Setting default max connections per route (100)
17/03/02 07:49:42 INFO authprovider.OAuth2AuthenticationProviderImpl: User was not cached or token did not match, thus querying the IdM
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: Doing request: GET https://account.lab.fiware.org/user?access_token=xxxxx HTTP/1.1
17/03/02 07:49:42 DEBUG authprovider.OAuth2AuthenticationProviderImpl: Response received: {xxxxx}
17/03/02 07:49:42 INFO authprovider.OAuth2AuthenticationProviderImpl: User frb authenticated
17/03/02 07:49:42 INFO authprovider.OAuth2AuthenticationProviderImpl: User cached
..
..
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: Hive configuration read
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: com.telefonica.iot.idm.endpoint=https://account.lab.fiware.org
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.enable.doAs=true
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.authentication=CUSTOM
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hive.server2.custom.authentication.class=com.telefonica.iot.cosmos.hive.authprovider.OAuth2AuthenticationProviderImpl
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hadoop.proxyuser.hive.groups=*
17/03/02 09:14:29 DEBUG authprovider.OAuth2AuthenticationProviderImpl: hadoop.proxyuser.hive.hosts=*
17/03/02 09:14:29 INFO authprovider.HttpClientFactory: Setting max total connections (500)
17/03/02 09:14:29 INFO authprovider.HttpClientFactory: Setting default max connections per route (100)
17/03/02 09:14:29 INFO authprovider.OAuth2AuthenticationProviderImpl: User and token were cached, thus nothing to query to IdM
```

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1 +1,2 @@
 - [cosmos-hive-auth-provider][hardening] Improve logs (#227)
+- [cosmos-hive-auth-provider][feature] Add OAuth2 cache (#225)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: No documentation associated to this PR?

File: cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java [hunk 2]
```diff
@@ -39,7 +39,8 @@
  */
 public class OAuth2AuthenticationProviderImpl implements PasswdAuthenticationProvider {
     
-    private static final Logger LOGGER = Logger.getLogger(HttpClientFactory.class);
+    private static final Logger LOGGER = Logger.getLogger(OAuth2AuthenticationProviderImpl.class);
+    private static final OAuth2Cache CACHE = new OAuth2Cache("/home/hive/oauth2.cache");
```
Comment (fgalan) [inline cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java]: It is ugly to have pathfile hardwired in source code... although maybe it doesn't worth the effort to put it outside, i.e. in configuration files.

Comment (frbattid) [inline cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java]: I thought about that while coding this. The problem is the only of way of maintaining an object (the cache) between different calls to the `OAuth2Cache` class is through a static final object like `CACHE`. And such kind of object can only be initialized when declared, i.e. cannot be initialized in the constructor (where a parameterized file path could be read, for sure).

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: Added in commit https://github.com/telefonicaid/fiware-cosmos/pull/230/commits/ed4debe7e4266e06a2e74faeb8fb089d2572dd11

File: cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java (hunk 2 above)
Comment (fgalan) [inline cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java]: Alternativelly, the filepath string could be moved to its own constant (in the .java where all constants live, if such place exists). It's still hardwired, but "softer" than current code ;)

File: doc/manuals/installation_and_administration_manual/oauth2_hive_auth_provider.md [hunk 3]
```diff
@@ -247,8 +250,28 @@ If the token exists but does not match the given user, then something like this
...
+```
+17/03/02 09:14:20 INFO authprovider.OAuth2AuthenticationProviderImpl: User was not cached or token did not match, thus querying the IdM
+...
+17/03/02 09:14:20 INFO authprovider.OAuth2AuthenticationProviderImpl: User cached
+...
+17/03/02 09:14:29 INFO authprovider.OAuth2AuthenticationProviderImpl: User and token were cached, thus nothing to query to IdM
+...
 ```
 
+With regards to the backup file for the cache, this is a file saving the cache between runs of HiveServer2, loaded each time the server starts, and saved each time a modification is done in the cache. It is always saved as `/home/hive/oauth2.cache` file.
```
Comment (fgalan) [inline doc/manuals/installation_and_administration_manual/oauth2_hive_auth_provider.md]: The following sentence could be added to remark this is not changeable:

"This pathfaile is not configurable (it is hardwired in the code)"

Comment (frbattid) [inline doc/manuals/installation_and_administration_manual/oauth2_hive_auth_provider.md]: Fixed in commit https://github.com/telefonicaid/fiware-cosmos/pull/230/commits/4d2b869ce22b98aa36407ea078891bd60cfbd938

File: cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java (hunk 2 above)
Comment (frbattid) [inline cosmos-hive-auth-provider/src/main/java/com/telefonica/iot/cosmos/hive/authprovider/OAuth2AuthenticationProviderImpl.java]: Fixed in commits https://github.com/telefonicaid/fiware-cosmos/pull/230/commits/5e575080d4ff7668e010c52db1f72f024e1ae7c5 and https://github.com/telefonicaid/fiware-cosmos/pull/230/commits/dfa819d2d4db9b21aa7547779cac775e9d69c09b

**Comment by fgalan** [comment]
LGTM

**Pull request merged**